File: VB/DevExpress.Calendar.Model/Appointment.vb
Imports Microsoft.VisualBasic
	Imports System
	Imports DevExpress.Xpo
	Imports DevExpress.XtraScheduler
	Imports DevExpress.XtraScheduler.Xml
	Imports DevExpress.Xpo.Services
	Imports DevExpress.Data.Services
Namespace DevExpress.Calendar.Model

#If (Not WIN) Then
    <ResourceSetName("Appointments")> _
Partial Public Class Appointment
        Inherits XPCustomObject
#Else
<Persistent("Appointments")> _
Partial Public Class Appointment
        Inherits XPCustomObject
#End If

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Public Sub New()
        End Sub

        Public Sub New(ByVal session As Session, ByVal classInfo As DevExpress.Xpo.Metadata.XPClassInfo)
            MyBase.New(session, classInfo)
        End Sub

        Public Overrides Sub AfterConstruction()
            MyBase.AfterConstruction()
        End Sub

        Protected Overrides Sub OnDeleted()
            MyBase.OnDeleted()
        End Sub

        Private _id As Guid
        <Key(AutoGenerate:=True)> _
        Public Property ID() As Guid
            Get
                Return _id
            End Get
            Set(ByVal value As Guid)
                SetPropertyValue(Of Guid)("ID", _id, value)
            End Set
        End Property

        Private _isAllDay As Boolean
        Public Property IsAllDay() As Boolean
            Get
                Return _isAllDay
            End Get
            Set(ByVal value As Boolean)
                SetPropertyValue(Of Boolean)("IsAllDay", _isAllDay, value)
            End Set
        End Property

        Private _description As String
        <Size(SizeAttribute.Unlimited)> _
        Public Property Description() As String
            Get
                Return _description
            End Get
            Set(ByVal value As String)
                SetPropertyValue(Of String)("Description", _description, value)
            End Set
        End Property

        Private _notes As String
        <Size(SizeAttribute.Unlimited)> _
        Public Property Notes() As String
            Get
                Return _notes
            End Get
            Set(ByVal value As String)
                SetPropertyValue(Of String)("Notes", _notes, value)
            End Set
        End Property

        Private _label As Integer
        Public Property Label() As Integer
            Get
                Return _label
            End Get
            Set(ByVal value As Integer)
                SetPropertyValue(Of Integer)("Label", _label, value)
            End Set
        End Property

        Private _location As String
        Public Property Location() As String
            Get
                Return _location
            End Get
            Set(ByVal value As String)
                SetPropertyValue(Of String)("Location", _location, value)
            End Set
        End Property

        Private _recurrence As String
        <Size(SizeAttribute.Unlimited)> _
        Public Property Recurrence() As String
            Get
                Return _recurrence
            End Get
            Set(ByVal value As String)
                SetPropertyValue(Of String)("Recurrence", _recurrence, value)
            End Set
        End Property

        Private _reminder As String
        <Size(SizeAttribute.Unlimited)> _
        Public Property Reminder() As String
            Get
                Return _reminder
            End Get
            Set(ByVal value As String)
                SetPropertyValue(Of String)("Reminder", _reminder, value)
            End Set
        End Property

        <NonPersistent()> _
        Public Property IsCompleted() As Boolean
            Get
                Return CompletionStatus < 0
            End Get
            Set(ByVal value As Boolean)
                CompletionStatus = If(value, CInt(Fix(CompletionCode.Completed)), CInt(Fix(CompletionCode.NotStarted)))
            End Set
        End Property

        Private _completionStatus As Integer
        Public Property CompletionStatus() As Integer
            Get
                Return _completionStatus
            End Get
            Set(ByVal value As Integer)
                SetPropertyValue(Of Integer)("CompletionStatus", _completionStatus, value)
            End Set
        End Property

        Private _start As DateTime
        Public Property Start() As DateTime
            Get
                Return _start
            End Get
            Set(ByVal value As DateTime)
                SetPropertyValue(Of DateTime)("Start", _start, value)
            End Set
        End Property

        Private _finish As DateTime
        Public Property Finish() As DateTime
            Get
                Return _finish
            End Get
            Set(ByVal value As DateTime)
                SetPropertyValue(Of DateTime)("Finish", _finish, value)
            End Set
        End Property

        Private _status As Integer
        Public Property Status() As Integer
            Get
                Return _status
            End Get
            Set(ByVal value As Integer)
                SetPropertyValue(Of Integer)("Status", _status, value)
            End Set
        End Property

        Private _subject As String
        Public Property Subject() As String
            Get
                Return _subject
            End Get
            Set(ByVal value As String)
                SetPropertyValue(Of String)("Subject", _subject, value)
            End Set
        End Property

        Private _itemType As Integer
        Public Property ItemType() As Integer
            Get
                Return _itemType
            End Get
            Set(ByVal value As Integer)
                SetPropertyValue(Of Integer)("ItemType", _itemType, value)
            End Set
        End Property

        Private _appointmentType As Integer
        Public Property AppointmentType() As Integer
            Get
                Return _appointmentType
            End Get
            Set(ByVal value As Integer)
                SetPropertyValue(Of Integer)("AppointmentType", _appointmentType, value)
            End Set
        End Property

#If RESOURCE_SHARING Then
		<Association("Appointments-Resources", UseAssociationNameAsIntermediateTableName := True)> _
		Public ReadOnly Property Resources() As XPCollection(Of Resource)
			Get
				Return GetCollection(Of Resource)("Resources")
			End Get
		End Property
#Else
        Private _resource As Resource
        <Association("Appointments-Resources", UseAssociationNameAsIntermediateTableName:=True)> _
        Public Property Resource() As Resource
            Get
                Return _resource
            End Get
            Set(ByVal value As Resource)
                SetPropertyValue(Of Resource)("Resource", _resource, value)
            End Set
        End Property

        <Visible(), NonPersistent()> _
        Public Property AssignedTo() As Guid?
            Get
                Return If(_resource IsNot Nothing, New Guid?(_resource.ID), Nothing)
            End Get
            Set(ByVal value? As Guid)
                Resource = Me.Session.GetObjectByKey(Of Resource)(value)
            End Set
        End Property
#End If

#If RESOURCES Then
#If (Not WIN) Then
		<NonPersistent(), Visible> _
		Public Property Resources() As String
#Else
		<NonPersistent()> _
		Public Property Resources() As String
#End If
			Get
'INSTANT VB NOTE: The local variable resources was renamed since Visual Basic will not allow local variables with the same name as their enclosing function or property:
				Dim resources_Renamed As New ResourceIdCollection()

				For i As Integer = 0 To Participants.Count - 1
					resources_Renamed.Add(Participants(i).ID)
				Next i

				Return New AppointmentResourceIdCollectionXmlPersistenceHelper(resources_Renamed).ToXml()
			End Get
			Set(ByVal value As String)
'INSTANT VB NOTE: The local variable resources was renamed since Visual Basic will not allow local variables with the same name as their enclosing function or property:
				Dim resources_Renamed As New ResourceIdCollection()

				If (Not String.IsNullOrEmpty(value)) Then
					resources_Renamed = AppointmentResourceIdCollectionXmlPersistenceHelper.ObjectFromXml(resources_Renamed, value)
				End If

				Participants.SuspendChangedEvents()
				Try
					Do While Participants.Count > 0
						Participants.Remove(Participants(0))
					Loop

					For i As Integer = 0 To resources_Renamed.Count - 1
						Dim resource As Participant = Me.Session.GetObjectByKey(Of Participant)(resources_Renamed(i))
						If resource IsNot Nothing Then
							Participants.Add(resource)
						End If
					Next i

				Finally
					Participants.ResumeChangedEvents()
				End Try
			End Set
		End Property
#End If

    End Class

End Namespace

File: CS/DevExpress.Calendar.Model/Appointment.cs
namespace DevExpress.Calendar.Model {
    using System;
    using DevExpress.Xpo;
    using DevExpress.XtraScheduler;
    using DevExpress.XtraScheduler.Xml;
#if !WIN
    using DevExpress.Xpo.Services;
    using DevExpress.Data.Services;
#endif

#if !WIN
    [ResourceSetName("Appointments")]
#endif
    [Persistent("Appointments")]
    public partial class Appointment : XPCustomObject {
        public Appointment(Session session) 
            : base(session) { 
        }

        public Appointment() {
        }

        public Appointment(Session session, DevExpress.Xpo.Metadata.XPClassInfo classInfo)
            : base(session, classInfo) {            
        }

        public override void AfterConstruction() {
            base.AfterConstruction();
        }

        protected override void OnDeleted() {
            base.OnDeleted();
        }
        
        Guid _id;
        [Key(AutoGenerate = true)]
        public Guid ID {
            get {
                return _id;
            }
            set {
                SetPropertyValue<Guid>("ID", ref _id, value);
            }
        }

        bool _isAllDay;
        public bool IsAllDay {
            get { 
                return _isAllDay; 
            }
            set { 
                SetPropertyValue<bool>("IsAllDay", ref _isAllDay, value); 
            }
        }

        string _description;
        [Size(SizeAttribute.Unlimited)]
        public string Description {
            get { 
                return _description; 
            }
            set { 
                SetPropertyValue<string>("Description", ref _description, value); 
            }
        }

        string _notes;
        [Size(SizeAttribute.Unlimited)]
        public string Notes {
            get {
                return _notes;
            }
            set {
                SetPropertyValue<string>("Notes", ref _notes, value);
            }
        }

        int _label;
        public int Label {
            get { 
                return _label; 
            }
            set { 
                SetPropertyValue<int>("Label", ref _label, value); 
            }
        }

        string _location;
        public string Location {
            get { 
                return _location; 
            }
            set {
                SetPropertyValue<string>("Location", ref _location, value); 
            }
        }

        string _recurrence;
        [Size(SizeAttribute.Unlimited)]
        public string Recurrence {
            get { 
                return _recurrence; 
            }
            set {
                SetPropertyValue<string>("Recurrence", ref _recurrence, value); 
            }
        }

        string _reminder;
        [Size(SizeAttribute.Unlimited)]
        public string Reminder {
            get { 
                return _reminder; 
            }
            set { 
                SetPropertyValue<string>("Reminder", ref _reminder, value); 
            }
        }

        [NonPersistent]
        public bool IsCompleted {
            get {
                return CompletionStatus < 0;
            }
            set {
                CompletionStatus = value ? 
                    (int)CompletionCode.Completed : (int)CompletionCode.NotStarted;
            }
        }

        int _completionStatus;
        public int CompletionStatus {
            get {
                return _completionStatus;
            }
            set {
                SetPropertyValue<int>("CompletionStatus", ref _completionStatus, value);
            }
        }

        DateTime _start;
        public DateTime Start {
            get { 
                return _start; 
            }
            set { 
                SetPropertyValue<DateTime>("Start", ref _start, value); 
            }
        }

        DateTime _finish;
        public DateTime Finish {
            get {
                return _finish;
            }
            set {
                SetPropertyValue<DateTime>("Finish", ref _finish, value);
            }
        }

        int _status;
        public int Status {
            get { 
                return _status; 
            }
            set { 
                SetPropertyValue<int>("Status", ref _status, value); 
            }
        }

        string _subject;
        public string Subject {
            get { 
                return _subject; 
            }
            set { 
                SetPropertyValue<string>("Subject", ref _subject, value); 
            }
        }

        int _itemType;
        public int ItemType {
            get {
                return _itemType;
            }
            set {
                SetPropertyValue<int>("ItemType", ref _itemType, value);
            }
        }

        int _appointmentType;
        public int AppointmentType {
            get { 
                return _appointmentType; 
            }
            set { 
                SetPropertyValue<int>("AppointmentType", ref _appointmentType, value); 
            }
        }

#if RESOURCE_SHARING
        [Association("Appointments-Resources", UseAssociationNameAsIntermediateTableName = true)]
        public XPCollection<Resource> Resources {
            get {
                return GetCollection<Resource>("Resources");
            }
        }
#else        
        Resource _resource;
        [Association("Appointments-Resources", UseAssociationNameAsIntermediateTableName = true)]
        public Resource Resource {
            get {
                return _resource;
            }
            set {
                SetPropertyValue<Resource>("Resource", ref _resource, value);
            }
        }

        [Visible]
        [NonPersistent]
        public Nullable<Guid> AssignedTo {
            get {
                return _resource != null 
                    ? new Nullable<Guid>(_resource.ID) : null;
            }
            set {
                Resource = this.Session.GetObjectByKey<Resource>(value);
            }
        }
#endif

#if RESOURCES
        [NonPersistent()]
#if !WIN
        [Visible]
#endif
        public string Resources {
            get {
                ResourceIdCollection resources = new ResourceIdCollection();
                
                for (int i = 0; i < Participants.Count; i++) {
                    resources.Add(Participants[i].ID);
                }

                return new AppointmentResourceIdCollectionXmlPersistenceHelper(resources).ToXml();
            }
            set {
                ResourceIdCollection resources = new ResourceIdCollection();
                
                if (!string.IsNullOrEmpty(value)) {
                    resources
                        = AppointmentResourceIdCollectionXmlPersistenceHelper.ObjectFromXml(resources, value);
                }
                
                Participants.SuspendChangedEvents();
                try {
                    while (Participants.Count > 0) {
                        Participants.Remove(Participants[0]);
                    }

                    for (int i = 0; i < resources.Count; i++) {
                        Participant resource = this.Session.GetObjectByKey<Participant>(resources[i]);
                        if (resource != null) {
                            Participants.Add(resource);
                        }
                    }

                } finally {
                    Participants.ResumeChangedEvents();
                }
            }
        }
#endif
         
    }

}